src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignerInfoTests.cs

```diff
@@ -731,6 +731,200 @@ public static void AddCounterSigner_ECDSA(SubjectIdentifierType identifierType,
+                    {
+                        IncludeOption = X509IncludeOption.None,
+                    });
+
+                CmsSigner counterSigner;
+
+                if (withCertificate)
+                {
+                    counterSigner = new CmsSigner(SubjectIdentifierType.NoSignature, cert);
+                }
+                else
+                {
+                    counterSigner = new CmsSigner(SubjectIdentifierType.NoSignature);
+                }
+
+                if (addExtraCert)
+                {
+                    counterSigner.Certificates.Add(cert2);
+                }
+
+                firstSigner.ComputeCounterSignature(counterSigner);
+
+                certs = cms.Certificates;
+
+                if (addExtraCert)
+                {
+                    Assert.Equal(2, certs.Count);
+                    Assert.NotEqual(cert2.RawData, certs[0].RawData);
+                    Assert.Equal(cert2.RawData, certs[1].RawData);
+                }
+                else
+                {
+                    Assert.Equal(1, certs.Count);
+                    Assert.NotEqual(cert2.RawData, certs[0].RawData);
+                }
+            }
+
+            Assert.ThrowsAny<CryptographicException>(() => cms.CheckSignature(true));
+            cms.CheckHash();
+
+            byte[] encoded = cms.Encode();
+            cms = new SignedCms();
+            cms.Decode(encoded);
+            Assert.ThrowsAny<CryptographicException>(() => cms.CheckSignature(true));
+            cms.CheckHash();
+
+            firstSigner = cms.SignerInfos[0];
+            firstSigner.CheckSignature(verifySignatureOnly: true);
+            Assert.ThrowsAny<CryptographicException>(() => firstSigner.CheckHash());
+
+            // The NoSignature CounterSigner sorts first.
+            SignerInfo firstCounterSigner = firstSigner.CounterSignerInfos[0];
+            Assert.Equal(SubjectIdentifierType.NoSignature, firstCounterSigner.SignerIdentifier.Type);
+            Assert.ThrowsAny<CryptographicException>(() => firstCounterSigner.CheckSignature(true));
+
+            if (PlatformDetection.IsFullFramework)
+            {
+                // NetFX's CheckHash only looks at top-level SignerInfos to find the
+                // crypt32 CMS signer ID, so it fails on any check from a countersigner.
+                Assert.ThrowsAny<CryptographicException>(() => firstCounterSigner.CheckHash());
+            }
+            else
+            {
+                firstCounterSigner.CheckHash();
+            }
+
+            certs = cms.Certificates;
+
+            if (addExtraCert)
+            {
+                Assert.Equal(2, certs.Count);
+                Assert.Equal("CN=DfHelleKeyAgreement1", certs[1].SubjectName.Name);
+            }
+            else
+            {
+                Assert.Equal(1, certs.Count);
+            }
+
+            Assert.Equal("CN=RSAKeyTransferCapi1", certs[0].SubjectName.Name);
+        }
+
```

Comment:
Is there a test case with: signer with private key, second one NoSignature no cert - what happens if you try to CheckSignature?

Reply:
The cert from the CmsSigner on NoSignature isn't used.  The places that "require" it checked the Certificate property for null without checking the signer type.  So there's not really a test benefit to AddSecond..._NoCertificate.

SignedCms.CheckSignature throws if any NoSignatures are present as signers or first-level counter-signers (we ignore second-level+ counter-signers).

SignerInfo.CheckSignature throws if the signer is NoSignature (and it ignores counter-signers).

Reply:
sounds good